RNA: In-Place Connection

sergeymitr opened the pull request:
#### Changes proposed in this Pull Request:
* RNA directory and testing environment.
* The Jetpack's AuthIframe component is partially extracted into the RNA component `InPlaceConnection`.
* Jetpack build and GUI testing script are adjusted to resolve `@automattic/jetpack-connection` to the RNA directory.

#### Jetpack product discussion
p1HpG7-bmK-p2
Partially extracted from the larger PR #18784.

#### Does this pull request change what data or activity we track or use?
No.

#### Testing instructions:
1. Build Jetpack using `yarn build`, confirm that it builds smoothly.
2. Go to `projects/plugins/jetpack/` and run `yarn test-gui`. Confirm that all tests pass.
3. Go to `projects/js-packages/connection` and run `yarn test`. Confirm that all 5 tests pass.
4. Connect Jetpack, create another WP admin account, and login using it.
5. Go to Jetpack Dashboard and click "Link to WordPress.com" button in the top banner. Confirm that the "Connect your WordPress.com account" iframe shows up.
6. Finish the secondary user connection process. Confirm that it went well.

sergeymitr committed:
RNA Connection: InPlaceConnection Component.

- The component itself.
- Unit tests.
- Replacing some Jetpack code with that new component.

sergeymitr committed:
RNA unit testing infrastructure.

sergeymitr committed:
Jetpack GUI testing: adjusting for linked RNA components.

anomiex commented on projects/plugins/jetpack/babel.config.js:
```diff
@@ -20,7 +20,17 @@ const config = {
 			presets: [ [ '@babel/preset-env', { targets: { node: 'current' } } ] ],
-			plugins: [ '@babel/plugin-transform-runtime' ],
+			plugins: [
+				'@babel/plugin-transform-runtime',
+				[
+					'module-resolver',
+					{
+						alias: {
+							'@automattic/jetpack-connection': '../../js-packages/connection',
+						},
+					},
+				],
```
Can we avoid this mess with dependencies in package.json? Either with a dep like `link:../../js-packages/connection` or using [yarn workspaces](https://classic.yarnpkg.com/en/docs/workspaces/) and a versioned dep?

sergeymitr replied:
Good idea, I replaced the resolver with a Yarn link.

anomiex commented on projects/plugins/jetpack/tests/runner.js:
```diff
@@ -1,7 +1,10 @@
-require( '@babel/register' );
+require( '@babel/register' )( {
+	ignore: [ /node_modules/ ],
+	only: [ '.', '../../js-packages/' ],
```
Same comment as projects/plugins/jetpack/babel.config.js?

sergeymitr replied:
We still need this, Yarn link doesn't solve this problem.
The `js-packages` directory is located outside the Babel root dir, so the package doesn't get built when tests run.
I know it's hacky, but gets the job donw.

This is temporary anyway, because once we move the Jetpack testing script into the repo root, `js-packages` will be inside the Babel root dir, and will get built by default.

anomiex commented on projects/plugins/jetpack/tools/webpack.config.js:
```diff
@@ -22,6 +22,12 @@ const sharedWebpackConfig = {
 	resolve: {
 		...baseWebpackConfig.resolve,
 		modules: [ path.resolve( path.dirname( __dirname ), '_inc/client' ), 'node_modules' ],
+		alias: {
+			'@automattic/jetpack-connection': path.resolve(
+				__dirname,
+				'../../../js-packages/connection'
+			),
+		},
```
Same comment as projects/plugins/jetpack/babel.config.js

anomiex reviewed (dismissed):
Most of the comments from https://github.com/Automattic/jetpack/pull/18784#pullrequestreview-588650555 still apply to this PR.

sergeymitr committed:
Replace Babel and Webpack resolvers with a Yarn link to the RNA Connection package.

sergeymitr committed:
RNA Connection: GitHub label handler.

sergeymitr committed:
Adding the changelog entry.

sergeymitr committed:
Adding the `sideEffects` entry into `package.json`.

sergeymitr committed:
Adjusting the package docs and comments.

sergeymitr committed:
Merge branch 'master' into add/rna-connection-in-place

sergeymitr committed:
Explicitly requiring `@babel/runtime/helpers/interopRequireDefault` in test runner.

Automated JS tests cannot load that module in some environments, so we explicitly load it by the test runner to avoid the problem.

sergeymitr committed:
Merge branch 'master' into add/rna-connection-in-place

sergeymitr committed:
Switching to the new generic test runner.

sergeymitr committed:
Adding the `composer.json` for `js-packages`.

anomiex commented on projects/js-packages/connection/composer.json:
```diff
@@ -0,0 +1,4 @@
+{
+	"name": "automattic/js-connection",
+	"description": "Jetpack Connection Component."
+}
```
I recommend adding changelogger right away. And you'll want your tests to be run by the CI infrastructure.
```suggestion
	"require": {},
	"require-dev": {
		"automattic/jetpack-changelogger": "1.1.x-dev"
	},
	"scripts": {
		"test-js": [
			"Composer\\Config::disableProcessTimeout",
			"yarn install",
			"yarn test"
		],
		"test-coverage": [
			"Composer\\Config::disableProcessTimeout",
			"yarn install",
			"whatever command will run the tests and produce coverage at $COVERAGE_DIR/clover.xml"
		]
	},
	"repositories": [
		{
			"type": "path",
			"url": "../../packages/*",
			"options": {
				"monorepo": true
			}
		}
	]
}
```
If you'll need a build step, that would go inside `scripts` too, like
```
		"build-development": [
			"Composer\\Config::disableProcessTimeout",
			"yarn build"
		]

```
You can have a "build-production" script too if there's a difference.

Once you've decided on a mirror repo, you'd also want to add
```
	"extra": {
		"mirror-repo": "Automattic/????",
		"changelogger": {
			"link-template": "https://github.com/Automattic/????/compare/v${old}...v${new}"
		}
	}
```
in there, with the `????` replaced of course.

anomiex commented on projects/js-packages/babel.config.js:
```diff
@@ -0,0 +1,11 @@
+const config = {
```
This still needs to be moved into `projects/js-packages/connection`.

anomiex commented on projects/js-packages/package.json:
```diff
@@ -0,0 +1,18 @@
+{
```
This needs to be merged into `projects/js-packages/connection`, we shouldn't have a package.json here.

anomiex commented on projects/js-packages/test-init.js:
```diff
@@ -0,0 +1,7 @@
+/**
```
This needs to go in `projects/js-packages/connection` too, or else find somewhere for it near `js-test-runner.js`.

sergeymitr committed:
Merge branch 'master' into add/rna-connection-in-place

sergeymitr committed:
Readjusting the test environment, adding coverage script.

sergeymitr committed:
Adding the changelog entry.

sergeymitr committed:
Adding the `CHANGELOG.md` file.

sergeymitr committed:
Upgrading the version to 0.1.0-alpha.

sergeymitr committed:
Test coverage configuration.

anomiex commented on projects/js-packages/connection/composer.json:
```diff
@@ -0,0 +1,29 @@
+		"test-coverage": [
+			"Composer\\Config::disableProcessTimeout",
+			"yarn install",
+			"yarn nyc --reporter=clover -x './!(node_modules)/**/test/*.jsx' --report-dir=\"$COVERAGE_DIR\" yarn test"
```
Looks like there's something wrong with the coverage, it's producing an empty coverage file (which then confuses CodeClimate).

anomiex reviewed (commented):
I only see two issues from the monorepo perspective at the moment now

* The first is pretty trivial: you should add an empty file at `projects/js-packages/connection/changelog/.gitkeep` so git won't go deleting the directory when a release happens.
* ~~The second it looks like you're already working on, for some reason `nyc` is outputting an empty coverage file.~~ Already fixed, but needed removing the nyc cache.

sergeymitr committed:
Add the .gitkeep file to the new changelog directory.

sergeymitr committed:
Add `js-packages` into the CLI project types.

fgiannar commented on projects/js-packages/connection/components/in-place-connection/index.jsx:
```diff
@@ -0,0 +1,128 @@
+				break;
+			case 'wpcom_nocookie':
+				// Third-party cookies blocked. Let's redirect.
+				window.location.replace( connectUrl );
```
How about we make this configurable for our generic component?
eg add an `onThirdPartyCookiesBlocked` prop?

sergeymitr replied:
Good point, I extracted this code into a helper: https://github.com/Automattic/jetpack/pull/19134/commits/3ec1cbf74745384a087627ef942259b4373cb17a

fgiannar commented on projects/js-packages/connection/components/in-place-connection/index.jsx:
```diff
@@ -0,0 +1,128 @@
+
+	// The URL looks like https://jetpack.wordpress.com/jetpack.authorize_iframe/1/. We need to include the trailing
+	// slash below so that we don't end up with something like /jetpack.authorize_iframe_iframe/
+	let src = connectUrl.replace( 'authorize/', 'authorize_iframe/' );
```
For **future** reference: I never really liked this.. I think the connectUrl (at least the base) should be a const.
We want a certain iframe, we expect certain events from it as well (eg `wpcom_nocookie`) so ideally we should be enforcing it.

Furthermore, the logic is currently a bit messed. Ideally this component should rely on information we can fetch from the connection package (independent from Jetpack), however we currently get this from `build_authorize_url` (defined in `class.jetpack.php`).

We don't need to change it now, just something that troubles me..

sergeymitr replied:
I agree, we need to fetch the required data from the connection package.
And I have started implementing this in another PR, so eventually the component will be more autonomous.
The `InPlaceConnection` component is a smaller part of it, and doesn't need to deal with fetching the data.

fgiannar commented on projects/js-packages/connection/components/in-place-connection/index.jsx:
```diff
@@ -0,0 +1,128 @@
+	// slash below so that we don't end up with something like /jetpack.authorize_iframe_iframe/
+	let src = connectUrl.replace( 'authorize/', 'authorize_iframe/' );
+
+	if ( hasConnectedOwner ) {
```
I wonder if it would be cleaner if we were more explicit here, aka rename `hasConnectedOwner` to `displayTos`?

sergeymitr replied:
Good idea.
I adjusted the behavior, and made it display TOS for everybody in userless mode: https://github.com/Automattic/jetpack/pull/19134/commits/8d5a044a1db73e1649b65a692c33190eba0d0076

fgiannar commented on projects/js-packages/connection/components/in-place-connection/index.jsx:
```diff
@@ -0,0 +1,128 @@
+ * @param {boolean} props.scrollToIframe -- Whether we need to auto-scroll the window upon element rendering.
+ * @param {string} props.connectUrl -- The connection URL.
+ * @param {Function} props.onComplete -- The callback to be called upon complete of the connection process.
+ * @param {string} props.source -- Component location identifier passed to WP.com.
```
At first glance I found this a bit confusing as I was expecting an override or something related to the iframe's source.
Do you think it would make sense to rename it to `triggeredBy` or something similar?

sergeymitr replied:
You're right, `source` is kind of ambiguous.
`triggeredBy` is better, but it assumes that the calling component is important, while we need the location in the UI, and don't care that much what exactly triggered it. We use the value to adjust the iframe styling, and it only cares about where the iframe will be displayed.
I don't have any better ideas though, so I'll need think about it.
If you have any thoughts, please share 🙂

fgiannar replied:
`location` maybe? We can come back to this later as it involves changes on WPCOM as well.

mdbitz replied:
based on the description "location identifier"
I would tend to agree with location

sergeymitr replied:
Renamed to `location` 🙂

fgiannar commented on projects/js-packages/connection/components/in-place-connection/index.jsx:
```diff
@@ -0,0 +1,128 @@
+	isLoading: false,
+	height: '220',
+	width: '100%',
+	scrollToIframe: true,
```
Maybe in the more generic context of this component we should default `scrollToIframe` to `false`?

sergeymitr replied:
Done 👍 
https://github.com/Automattic/jetpack/pull/19134/commits/6b5997ee470284018ee78c53e087180f804f8844

fgiannar commented on projects/js-packages/connection/components/in-place-connection/index.jsx:
```diff
@@ -0,0 +1,128 @@
+	let src = connectUrl.replace( 'authorize/', 'authorize_iframe/' );
+
+	if ( hasConnectedOwner ) {
+		src += '&display-tos';
```
Are we 100% certain that the url will always have existing params? aka `?` vs `&`
Is it an overkill to handle this here?

sergeymitr replied:
Good catch!
Although the URL always contains GET-parameters, we shouldn't rely on this to be true.
https://github.com/Automattic/jetpack/pull/19134/commits/45dd5e79328ebf43706c832372cdfdcd00c25219

fgiannar reviewed (commented):
Hey Sergey!

Great work here! I've left some comments, mostly to discuss a few concerns, nothing is a blocker just with future maintainability/flexibility in mind.

Looking forward to your thoughts! Thanks for working on this!

sergeymitr committed:
Extracting the third-party cookies fallback into a helper function.

sergeymitr committed:
Docs for the `onThirdPartyCookiesBlocked` property.

sergeymitr committed:
TOS for InPlaceConnection component.

- new `isSiteRegistered()` reducer function to check if the site is registered (has blog token)
- renaming the `hasConnectedOwner` property to `displayTOS`
- display TOS not only for secondary users, but also for all users in userless mode

sergeymitr committed:
Make `scrollToIframe` default to `false`.

sergeymitr committed:
Add `?` to the iframe URL if it's missing.

sergeymitr committed:
Adjusting the unit tests according to recent changes.

sergeymitr committed:
Merge branch 'master' into add/rna-connection-in-place

sergeymitr committed:
Renamed the 3rd-party-cookie fallback helper file.

fgiannar commented on projects/js-packages/connection/README.md:
```diff
@@ -0,0 +1,41 @@
+	width={ this.props.width}
+	isLoading={ this.props.fetchingConnectUrl }
+	title={ this.props.title }
+	hasConnectedOwner={ this.props.hasConnectedOwner }
```
```suggestion
	displayTOS={ false }
```

Also (non trivial) we could maybe replace `this.props` logic here with actual example values to make it easier to understand.

Finally (again non trivial) we could also mention the nice fallback helper for 3rd party cookies you added!

sergeymitr replied:
You're right, we need the docs!
Done!

fgiannar reviewed (dismissed):
Hey Sergey!

Thank you for addressing all my comments!

I left 2 very minor ones (a typo and a suggestion for improving the README).

I tested using the provided instructions and all worked as expected! I also tested this with Safari to verify the redirect fallback for 3rd party cookies works :)

The only think that didn't work for me when running the tests was:
> Go to `projects/js-packages/` and run `yarn test`. Confirm that all 5 tests pass.

I had to go to `projects/js-packages/connection` instead.

Great work here! 🥇

anomiex committed:
Merge remote-tracking branch 'origin/master' into add/rna-connection-in-place

sergeymitr committed:
Update projects/js-packages/connection/README.md

Co-authored-by: Foteini Giannaropoulou <giannaropoulou.foteini@gmail.com>

sergeymitr committed:
Improved documentation.

jeherve commented on .github/actions/repo-gardening/src/tasks/add-labels/index.js:
```diff
@@ -125,6 +126,14 @@ async function getLabelsToAdd( octokit, owner, repo, number ) {
 			if ( project.groups.ptype === 'github-actions' ) {
 				keywords.add( 'Actions' );
 			}
+
+			if ( project.groups.ptype === 'packages' ) {
+				keywords.add( '[Status] Needs Package Release' );
+			}
+			if ( project.groups.ptype === 'js-packages' ) {
+				keywords.add( '[Status] Needs JS Package Release' );
+				keywords.add( 'RNA' );
+			}
```
We do not really use those labels anymore, now that we've automated most of the package. 

```suggestion
```

sergeymitr replied:
Interesting, I didn't know that.
Removing 👍

jeherve commented on projects/plugins/jetpack/_inc/client/components/auth-iframe/index.jsx:
```diff
@@ -6,6 +6,8 @@ import { noop } from 'lodash';
 import React from 'react';
 import { __ } from '@wordpress/i18n';
+import { InPlaceConnection } from '@automattic/jetpack-connection';
+import { thirdPartyCookiesFallbackHelper } from '@automattic/jetpack-connection/helpers';
```
Should we export `thirdPartyCookiesFallbackHelper` directly from `projects/js-packages/connection/index.jsx`, so we wouldn't have to rely on importing from a subfolder?

sergeymitr replied:
I've been asking myself the same question, actually.
By exporting helpers from a separate file we logically separate them from the actual components.
On the other hand, that might be excessive, and having "Helper" in the name seems to be enough.

sergeymitr committed:
Remove auto-assigning of "Needs [JS] Package Release" labels.

Co-authored-by: Jeremy Herve <jeremy@jeremy.hu>

mdbitz commented on projects/js-packages/connection/components/in-place-connection/index.jsx:
```diff
@@ -0,0 +1,137 @@
+	const iframeWrapRef = useRef();
+	const iframeRef = useRef();
+
+	const receiveData = e => {
```
We should document functions in the js files.

mdbitz commented on projects/plugins/jetpack/_inc/client/components/auth-iframe/test/component.js:
```diff
@@ -13,20 +12,21 @@ import { AuthIframe } from '../index';
-        scrollToIframe    : false,
+        connectUrl: 'https://jetpack.wordpress.com/jetpack.authorize/1/',
+        scrollToIframe: false,
+		hasConnectedOwner: false,
```
we should fix the indentation.

sergeymitr committed:
Merge branch 'master' into add/rna-connection-in-place

sergeymitr committed:
Move helpers from separate export file to `index.jsx`.

sergeymitr committed:
JsDoc and indentation.

sergeymitr committed:
Renamed property `source` to `location`.

sergeymitr committed:
Merge branch 'master' into add/rna-connection-in-place

zinigor reviewed (dismissed):
Thanks for this! Let's get the PR in once the tests pass.

sergeymitr committed:
Repo-gardening changelog.

zinigor reviewed (dismissed):
I still stand by my words! :)

sergeymitr committed:
Add Git attributes config.

jeherve commented:
@sergeymitr Could you commit D58560-code to close the loop on this?

Thank you!

jeherve commented:
r224062-wpcom
